internal/controller/ipam_controller.go

```diff
@@ -0,0 +1,204 @@
+	ConfigMapNamespace string
+	ConfigMapName      string
+	CIDRPoolsNamespace string
+}
+
+//+kubebuilder:rbac:groups="",resources=configmaps/finalizers,verbs=update
+//+kubebuilder:rbac:groups=nv-ipam.nvidia.com,resources=cidrpools,verbs=get;list;patch;delete;deletecollection
+
+func (r *NvIPAMReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logr := log.FromContext(ctx)
+	logr.Info("Reconciling Spectrum-X ConfigMap")
+	// Fetch the ConfigMap
+	cm := &corev1.ConfigMap{}
+	err := r.Get(ctx, req.NamespacedName, cm)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			// Return early if the object is not found.
+			return ctrl.Result{}, nil
+		}
+	}
+	// Handle deletion reconciliation loop.
+	if !cm.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, cm)
+	}
+
+	spectrumXConfig, err := config.ParseConfig(cm.Data[config.ConfigMapKey])
+	if err != nil {
+		logr.Info("Failed to parse Spectrum-X ConfigMap Data")
+		return ctrl.Result{}, err
+	}
+	logr.Info("Parsed Spectrum-X Config", "config", spectrumXConfig)
+
+	err = r.reconcileCIDRPools(ctx, spectrumXConfig)
+	if err != nil {
+		logr.Info("Failed to reconcile CIDRPool")
+		return ctrl.Result{}, err
+	}
+	// Add finalizer if not set.
+	if !controllerutil.ContainsFinalizer(cm, spectrumXFinalizer) {
+		logr.Info("Adding finalizer to configmap")
+		controllerutil.AddFinalizer(cm, spectrumXFinalizer)
+		cm.ObjectMeta.ManagedFields = nil
+		if err := r.Client.Patch(ctx, cm, client.Apply, client.ForceOwnership, client.FieldOwner(spectrumXIPAMControllerName)); err != nil {
+			return ctrl.Result{}, fmt.Errorf("error while adding finalizer to configmap %w", err)
+		}
+		return ctrl.Result{}, nil
+	}
+	return ctrl.Result{}, nil
+}
+
+// reconcileCIDRPools generates CIDRPools according to defined Rails
+func (r *NvIPAMReconciler) reconcileCIDRPools(ctx context.Context, cfg *config.Config) error {
+	for _, rail := range cfg.SpectrumXNetworks.Rails {
+		pool := r.generateCIDRPool(&rail, cfg.Hosts, cfg.SpectrumXNetworks.CrossRailSubnet)
+		if err := r.Client.Patch(ctx, pool, client.Apply, client.ForceOwnership, client.FieldOwner(spectrumXIPAMControllerName)); err != nil {
+			return fmt.Errorf("error while patching %s %s: %w", pool.GetObjectKind().GroupVersionKind().String(), client.ObjectKeyFromObject(pool), err)
+		}
+	}
+	return nil
+}
+
+// generateCIDRPool generates a CIDRPool object for the given Rail
+func (r *NvIPAMReconciler) generateCIDRPool(rail *config.Rail, hosts []config.Host, crossRailNet string) *nvipamv1.CIDRPool {
+	staticAllocations := getHostsAllocation(rail.Name, hosts)
+	pool := &nvipamv1.CIDRPool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rail.Name,
+			Namespace: r.CIDRPoolsNamespace,
+			Labels: map[string]string{
+				spectrumXIPAMControllerName: "",
+			},
+		},
+		Spec: nvipamv1.CIDRPoolSpec{
+			CIDR:                 rail.Subnet,
+			GatewayIndex:         ptr.To[int32](0),
+			PerNodeNetworkPrefix: 31,
```

Comment:
you should take if from the config as well no?

Reply:
It is always point to point